src/config/field-overrides/relationshipType.js

```diff
@@ -59,54 +67,127 @@ const styles = {
+                relationshipEntity: "TRACKED_ENTITY_INSTANCE",
+                trackedEntityType: { id }
+
+        When Program Stage Instance is selected, we need to know the programType for selected programs, so we keep a reference to the model.
+        We use the program-dropdown as a 'filter' (for trackerprograms) to only show programStages for the selected program,
+        however we cannot post this value to the server together with a programStage, as that results in an error.
+        We post just the programStageID for Tracker-programs, and just the programID for event-programs
+        Therefore we save the selected values in the state. */
+
+        let selected = null;
+        let relationshipEntity = props.value.relationshipEntity;
+        if (props.value && relationshipEntity) {
+            const modelTypes =
+                modelTypesForRelationshipEntity[relationshipEntity];
+            selected = {};
+            modelTypes.forEach(objOpts => {
+                const modelType = objOpts.modelType;
+                if (props.value[modelType]) {
+                    selected[modelType] = props.value[modelType];
+                }
+            });
         }
 
         this.state = {
-            relationshipEntity,
-            embeddedValue,
+            selected,
+            loading: true,
+            error: false,
         };
     }
 
-    selectRelationshipEntity = (_, __, value) => {
+    componentDidMount() {
+        /* Since we cannot get the programId from the same request as when loading the model,
+            (due to the nature of the dynamic structure of the relationShipConstraint object)
+            we need to grab the programStage with another request, as we need this to show the
+            program that the selected program stage belongs to. */
+        const selectedPS = this.state.selected && this.state.selected.programStage;
+        if(selectedPS) {
+            this.fetchProgramForProgramStage(selectedPS.id).then(programModel => {
+                this.setState(state => ({
+                    selected: {
+                        ...state.selected,
+                        program: programModel,
+                   },
+                   loading: false
+                }));
+            }).catch(e => {
+                this.setState({
+                    error: e
+                });
+            });
+        }
+        this.setState({loading: false});
+    }
+
+    getSelectedRelationshipEntity = () => {
+        const props = this.props;
+        return (props.value && props.value.relationshipEntity) || null;
+    };
+
+    getSelectedIDForModelType = modelType => {
+        // We use state here, since program and programStage cannot be selected
+        // at the same time (for the actual posted value). See constructor comment
+        return (this.state.selected && this.state.selected[modelType]) || null;
+    };
+
+    fetchProgramForProgramStage = async (programStageId) => {
```

Comment:
I think this can just be a regular method, you don't have to bind it to the this scope:
`async fetchProgramForProgramStage(programStageId) {`